tests/test_admin_views.py

```diff
@@ -0,0 +1,146 @@
+from django.test import Client, TestCase
+from django.contrib.auth.models import User
+from django.urls import reverse
+from tests import factories
+import pytest
+from unittest import mock
+
+
+@pytest.mark.django_db
+class TestSubjectAdd(TestCase):
+
+    def setUp(self):
+        self.admin_user = User.objects.create_superuser(
+                          'test_user', 'test_user@test.com', 'password')
+        self.admin_user.save()
+        self.c = Client()
+        self.c.force_login(self.admin_user)
+
+    def test_subject_add(self):
+        """Test main add menu is displayed."""
+        response = self.c.get(reverse('admin_main_add'))
+        assert response.status_code == 200
+        # Assert the subject will be added to the top level
+        assert response.context['browse_len'] == 0
+        assert response.context['saved'] is None
+        assert response.templates[0].name == 'subjects/add_subject.html'
+
+    def test_subject_add_to_sub_id(self):
+        """Test the AddSubject form for new subject will add it as a sub-category
+         to the current subject."""
+        subject = factories.SubjectFactory.create()
+        response = self.c.get(reverse('admin_add', kwargs={'sub_id': subject.id}))
+        # Assert the new subject will be added as a sub category of 'subject'.
+        assert response.context['browse_len'] == 1
+        assert response.context['browse_string'] == subject.name
+        assert response.context['saved'] is None
+        assert response.status_code == 200
+        assert response.templates[0].name == 'subjects/add_subject.html'
+
+    def test_subject_add_post_with_sub_id(self):
+        """Test the new subject is added."""
+        subject = factories.SubjectFactory.create()
+        subject_2 = factories.SubjectFactory.create()
+        with mock.patch('subjects.admin_views.AddSubject') as mock_form:
+            mock_form.return_value.is_valid.return_value = True
+            mock_form.return_value.cleaned_data = {'name': subject_2.name,
+                                                   'keywords': subject_2.keywords,
+                                                   'notes': subject_2.notes}
+            response = self.c.post(reverse('admin_add', kwargs={'sub_id': subject.id}))
```

Comment:
I am not sure about the design of this test where it looks to me we are creating `subject_2` in the database explicitly in the test then mocking the form we would have used to add `subject_2`. I wonder if instead we should not do these things and create the object by adding `data` to the POST request, like:
```
        name = 'boo'
        response = self.c.post(reverse('admin_add', kwargs={'sub_id': subject.id}),
             data={'name': name, 'keywords': 'blah blah', 'notes': 'some notes'})

        assert response.context['saved'].name == name
        assert response.status_code == 200
        assert response.templates[0].name == 'subjects/add_subject.html'
```
and verifying the object is created.